In [37]:
#import dependencies
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.linear_model import LogisticRegression

pd.set_option('display.max_columns', None)

In [45]:
#load in the data
application_df = pd.read_csv('../csv/application_record.csv')
credit_record_df = pd.read_csv("../csv/credit_record.csv")

In [46]:
application_df.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [47]:
credit_record_df.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [48]:
credit_df = pd.merge(application_df, credit_record_df, on='ID', how='inner')

In [49]:
credit_df['STATUS'].value_counts()

STATUS
C    329536
0    290654
X    145950
1      8747
5      1527
2       801
3       286
4       214
Name: count, dtype: int64

In [50]:
credit_df.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008804           M            Y               Y             0   
2  5008804           M            Y               Y             0   
3  5008804           M            Y               Y             0   
4  5008804           M            Y               Y             0   

   AMT_INCOME_TOTAL NAME_INCOME_TYPE NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  \
0          427500.0          Working    Higher education     Civil marriage   
1          427500.0          Working    Higher education     Civil marriage   
2          427500.0          Working    Higher education     Civil marriage   
3          427500.0          Working    Higher education     Civil marriage   
4          427500.0          Working    Higher education     Civil marriage   

  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  \
0  Rented apartment      -12005          -4542           1                1   
1  Rented apartment      -12005          -4542           1                1   
2  Rented apartment      -12005          -4542           1                1   
3  Rented apartment      -12005          -4542           1                1   
4  Rented apartment      -12005          -4542           1                1   

   FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  MONTHS_BALANCE  \
0           0           0             NaN              2.0               0   
1           0           0             NaN              2.0              -1   
2           0           0             NaN              2.0              -2   
3           0           0             NaN              2.0              -3   
4           0           0             NaN              2.0              -4   

  STATUS  
0      C  
1      C  
2      C  
3      C  
4      C

In [51]:
# 0: 1-29 days past due 
# 1: 30-59 days past due 
# 2: 60-89 days overdue 
# 3: 90-119 days overdue 
# 4: 120-149 days overdue 
# 5: Overdue or bad debts, write-offs for more than 150 days 
# C: paid off that month 
# X: No loan for the month

In [52]:
credit_df['target'] = credit_df['STATUS'].apply(lambda x: 0 if x in ['0','1','2', '3', '4', '5'] else 1)

In [53]:
credit_df.to_csv('../csv/application_credit_record_w_target.csv', index=False)

In [54]:
credit_df.drop(columns=['ID'], inplace=True)
credit_df.head()

CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  \
0           M            Y               Y             0          427500.0   
1           M            Y               Y             0          427500.0   
2           M            Y               Y             0          427500.0   
3           M            Y               Y             0          427500.0   
4           M            Y               Y             0          427500.0   

  NAME_INCOME_TYPE NAME_EDUCATION_TYPE NAME_FAMILY_STATUS NAME_HOUSING_TYPE  \
0          Working    Higher education     Civil marriage  Rented apartment   
1          Working    Higher education     Civil marriage  Rented apartment   
2          Working    Higher education     Civil marriage  Rented apartment   
3          Working    Higher education     Civil marriage  Rented apartment   
4          Working    Higher education     Civil marriage  Rented apartment   

   DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  \
0      -12005          -4542           1                1           0   
1      -12005          -4542           1                1           0   
2      -12005          -4542           1                1           0   
3      -12005          -4542           1                1           0   
4      -12005          -4542           1                1           0   

   FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  MONTHS_BALANCE STATUS  target  
0           0             NaN              2.0               0      C       1  
1           0             NaN              2.0              -1      C       1  
2           0             NaN              2.0              -2      C       1  
3           0             NaN              2.0              -3      C       1  
4           0             NaN              2.0              -4      C       1

begin model training

In [55]:
df = credit_df.copy()

df = pd.get_dummies(df, columns=['STATUS', 'CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE'], drop_first=True)

df.head()

CNT_CHILDREN  AMT_INCOME_TOTAL  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  \
0             0          427500.0      -12005          -4542           1   
1             0          427500.0      -12005          -4542           1   
2             0          427500.0      -12005          -4542           1   
3             0          427500.0      -12005          -4542           1   
4             0          427500.0      -12005          -4542           1   

   FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL  CNT_FAM_MEMBERS  MONTHS_BALANCE  \
0                1           0           0              2.0               0   
1                1           0           0              2.0              -1   
2                1           0           0              2.0              -2   
3                1           0           0              2.0              -3   
4                1           0           0              2.0              -4   

   target  STATUS_1  STATUS_2  STATUS_3  STATUS_4  STATUS_5  STATUS_C  \
0       1     False     False     False     False     False      True   
1       1     False     False     False     False     False      True   
2       1     False     False     False     False     False      True   
3       1     False     False     False     False     False      True   
4       1     False     False     False     False     False      True   

   STATUS_X  CODE_GENDER_M  FLAG_OWN_CAR_Y  FLAG_OWN_REALTY_Y  \
0     False           True            True               True   
1     False           True            True               True   
2     False           True            True               True   
3     False           True            True               True   
4     False           True            True               True   

   NAME_INCOME_TYPE_Pensioner  NAME_INCOME_TYPE_State servant  \
0                       False                           False   
1                       False                           False   
2                       False                           False   
3                       False                           False   
4                       False                           False   

   NAME_INCOME_TYPE_Student  NAME_INCOME_TYPE_Working  \
0                     False                      True   
1                     False                      True   
2                     False                      True   
3                     False                      True   
4                     False                      True   

   NAME_EDUCATION_TYPE_Higher education  \
0                                  True   
1                                  True   
2                                  True   
3                                  True   
4                                  True   

   NAME_EDUCATION_TYPE_Incomplete higher  NAME_EDUCATION_TYPE_Lower secondary  \
0                                  False                                False   
1                                  False                                False   
2                                  False                                False   
3                                  False                                False   
4                                  False                                False   

   NAME_EDUCATION_TYPE_Secondary / secondary special  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   

   NAME_FAMILY_STATUS_Married  NAME_FAMILY_STATUS_Separated  \
0                       False                         False   
1                       False                         False   
2                       False                         False   
3                       False                         False   
4                       False                         False   

   NAME_FAMILY_STATUS_Single / not married  N

In [59]:
y = df['target'].drop(columns=['ID'])

X = df.drop(columns=['target'])

In [60]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [61]:
X.head()

CNT_CHILDREN  AMT_INCOME_TOTAL  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  \
0             0          427500.0      -12005          -4542           1   
1             0          427500.0      -12005          -4542           1   
2             0          427500.0      -12005          -4542           1   
3             0          427500.0      -12005          -4542           1   
4             0          427500.0      -12005          -4542           1   

   FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL  CNT_FAM_MEMBERS  MONTHS_BALANCE  \
0                1           0           0              2.0               0   
1                1           0           0              2.0              -1   
2                1           0           0              2.0              -2   
3                1           0           0              2.0              -3   
4                1           0           0              2.0              -4   

   STATUS_1  STATUS_2  STATUS_3  STATUS_4  STATUS_5  STATUS_C  STATUS_X  \
0     False     False     False     False     False      True     False   
1     False     False     False     False     False      True     False   
2     False     False     False     False     False      True     False   
3     False     False     False     False     False      True     False   
4     False     False     False     False     False      True     False   

   CODE_GENDER_M  FLAG_OWN_CAR_Y  FLAG_OWN_REALTY_Y  \
0           True            True               True   
1           True            True               True   
2           True            True               True   
3           True            True               True   
4           True            True               True   

   NAME_INCOME_TYPE_Pensioner  NAME_INCOME_TYPE_State servant  \
0                       False                           False   
1                       False                           False   
2                       False                           False   
3                       False                           False   
4                       False                           False   

   NAME_INCOME_TYPE_Student  NAME_INCOME_TYPE_Working  \
0                     False                      True   
1                     False                      True   
2                     False                      True   
3                     False                      True   
4                     False                      True   

   NAME_EDUCATION_TYPE_Higher education  \
0                                  True   
1                                  True   
2                                  True   
3                                  True   
4                                  True   

   NAME_EDUCATION_TYPE_Incomplete higher  NAME_EDUCATION_TYPE_Lower secondary  \
0                                  False                                False   
1                                  False                                False   
2                                  False                                False   
3                                  False                                False   
4                                  False                                False   

   NAME_EDUCATION_TYPE_Secondary / secondary special  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   

   NAME_FAMILY_STATUS_Married  NAME_FAMILY_STATUS_Separated  \
0                       False                         False   
1                       False                         False   
2                       False                         False   
3                       False                         False   
4                       False                         False   

   NAME_FAMILY_STATUS_Single / not married  NAME_FAMILY_STATUS_Widow  \
0                    

train_test_split

In [62]:
x_train, x_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

Linear Regression Model

In [63]:
LogRegModel = LogisticRegression(random_state=1)

LogRegModel.fit(x_train, y_train)

/opt/anaconda3/envs/dev/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=1)

In [64]:
test_predictions = LogRegModel.predict(x_test)

In [67]:
LogRegCM = confusion_matrix(y_test, test_predictions)

df_LogRegDM = pd.DataFrame(
    LogRegCM,
    columns=['Predicted 0', 'Predicted 1'],
    index=['Actual 0', 'Actual 1']
)

df_LogRegDM.head()

Predicted 0  Predicted 1
Actual 0        62369        13188
Actual 1         9494       109378

In [68]:
LogRegCR = classification_report(y_test, test_predictions)

print('Confusion Matrix')
display(df_LogRegDM)
print('Classification Report')
print(LogRegCR)

Confusion Matrix


Predicted 0  Predicted 1
Actual 0        62369        13188
Actual 1         9494       109378

Classification Report
              precision    recall  f1-score   support

           0       0.87      0.83      0.85     75557
           1       0.89      0.92      0.91    118872

    accuracy                           0.88    194429
   macro avg       0.88      0.87      0.88    194429
weighted avg       0.88      0.88      0.88    194429



The logistic regression model does a good job predicting both 0 (bad) consumers and 1 (good) consumers.

Good Consumers (1): The model is 83% accurate here. 
Bad Consumers (0): The model is 92% accurate here

The model correctly identifies 92% of all good consumers. Out of the consumers it predicts as bad, 87% are actually bad. It balances these two measures well, with an overall score of 88% for bad consumers. Overall, the model is 88% accurate, and it handles both consumer types well. There is room for improvement to train the model to achieve a higher accuracy score.

# Neural Network
